### PR TITLE
Kodi Integration & Complete UdpApi Rework

### LINKED ISSUE #9: Feature request:  Configuration option to terminate (exit) application when banned
@Tyderion First I would like to thank you for providing this CLI version of aniadd, as well as working to improve it.

One annoying thing about the current implementation is that when you invariably get banned from the API, the client continues to check (which I believe just elongates your ban).

Would it be something you would be willing to add as an configuration option that when Banned from the API, the application terminates (Exit)?

I suggest a configuration option (default false) as that would be a good way to add the functionality without changing the application's default behavior.

### PR DESCRIPTION
**Pull request by Tyderion**

### CODE REVIEW

File: src/main/java/fileprocessor/FindFiles.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+        log.fine(STR."Folder: \{folder.getAbsolutePath()}");
+
+        val files = recursivelyGetFilesToScrape(folder).collect(Collectors.toList());
+        files.forEach(f -> log.finest(STR."Found file: \{f.getAbsolutePath()}"));
+        log.info(STR."Number of found files: \{files.size()}");
+        return files;
+    }
+
+    private Stream<File> recursivelyGetFilesToScrape(File folder) {
```
Comment (runecalico) [inline src/main/java/fileprocessor/FindFiles.java:33]: I *assume* this is supposed to return *only* files, and not the folders those files are on. However at least on Windows (and accessing an SMB share), it returns the folders, and then the files in them.

```
2024-07-21 09:23:25 FINE fileprocessor.FindFiles call Folder: Z:\xxxx\anidbclitest1
2024-07-21 09:23:25 FINEST fileprocessor.FindFiles lambda$call$0 Found file: Z:\xxxx\anidbclitest1\A Silent Voice [anidb-10937]
2024-07-21 09:23:25 FINEST fileprocessor.FindFiles lambda$call$0 Found file: Z:\xxxx\anidbclitest1\Adesugata Mahou no Sannin Musume [anidb-3633]
2024-07-21 09:23:25 FINEST fileprocessor.FindFiles lambda$call$0 Found file: Z:\xxxx\anidbclitest1\Akikan! (2009) [anidb-7132]
2024-07-21 09:23:25 FINEST fileprocessor.FindFiles lambda$call$0 Found file: Z:\xxxx\anidbclitest1\Akuma-kun Youkoso Akuma Land e!! [anidb-2051]
2024-07-21 09:23:25 FINEST fileprocessor.FindFiles lambda$call$0 Found file: Z:\xxxx\anidbclitest1\A Silent Voice [anidb-10937]\A Silent Voice [BlurayDesuYo][BluRay][1920x1080][HEVC][7273A66D].mkv
2024-07-21 09:23:25 FINEST fileprocessor.FindFiles lambda$call$0 Found file: Z:\xxxx\anidbclitest1\A Silent Voice [anidb-10937]\A Silent Voice [Okay-Subs][BluRay][1920x1036][HEVC][6457E1C3].mkv
2024-07-21 09:23:25 FINEST fileprocessor.FindFiles lambda$call$0 Found file: Z:\xxxx\anidbclitest1\Adesugata Mahou no Sannin Musume [anidb-3633]\Adesugata Mahou no Sannin Musume [raw][unknown][320x240][MS MP4x][65E7ABA9].avi
2024-07-21 09:23:25 FINEST fileprocessor.FindFiles lambda$call$0 Found file: Z:\xxxx\anidbclitest1\Akikan! (2009) [anidb-7132]\Akikan! (2009) [TSSMA][DVD][848x480][h264][D4784111].mkv
2024-07-21 09:23:25 FINEST fileprocessor.FindFiles lambda$call$0 Found file: Z:\xxxx\anidbclitest1\Akuma-kun Youkoso Akuma Land e!! [anidb-2051]\Akuma-kun Youkoso Akuma Land e!! [Rododendron][www][1280x720][h264][648B02E2].mkv
2024-07-21 09:23:25 INFO fileprocessor.FindFiles call Number of found files: 9
```
Which then results in an fatal exception when it tries to hash the directory
```
2024-07-21 09:23:25 FINE processing.Mod_EpProcessing addFiles File Count changed to 9
2024-07-21 09:23:25 INFO processing.Mod_EpProcessing start Starting processing
2024-07-21 09:23:25 FINE processing.Mod_EpProcessing processEps Processing file Z:\xxxx\anidbclitest1\A Silent Voice [anidb-10937] with Id 0
java.io.FileNotFoundException: Z:\xxx\anidbclitest1\A Silent Voice [anidb-10937] (Access is denied)
        at java.base/java.io.FileInputStream.open0(Native Method)
        at java.base/java.io.FileInputStream.open(FileInputStream.java:213)
        at java.base/java.io.FileInputStream.<init>(FileInputStream.java:152)
        at processing.FileParser.run(FileParser.java:25)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
```

File: src/main/java/fileprocessor/FindFiles.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    private final String directory;
+
+    @Override
+    public List<File> call() throws Exception {
+        File folder = new File(directory);
+
+        log.fine(STR."Folder: \{folder.getAbsolutePath()}");
+
+        val files = recursivelyGetFilesToScrape(folder).collect(Collectors.toList());
```
Comment (runecalico) [inline src/main/java/fileprocessor/FindFiles.java:27]: I couldn't figure out how to modify recursivelyGetFilesToScrape to return only files, but adding a filter to this worked enough to get me past the error.
```
        val files = recursivelyGetFilesToScrape(folder).filter(File::isFile).collect(Collectors.toList());
```

File: src/main/java/udpapi2/UdpApi.java [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+                queueLogin();
+            }
+            queueCommand(command);
+            return;
+        }
+        scheduleCommand(command, getNextSendDelay());
+    }
+
+    private void scheduleCommand(@NotNull Command command, Duration delay) {
+        logger.info(STR."Scheduling command \{command.toString()} in \{delay.toMillis()} ms");
```
Comment (runecalico) [inline src/main/java/udpapi2/UdpApi.java]: Might I suggest sanitizing the command output prior to logging it to mask the user/pass fields?
While the following code worked for me, I do not claim it is a good way to do this, and I would expect additional log locations in the udpApi class probably need to sanitize their output as well.
```
    private String sanitizedLogString(String message) {
        String sanitizedString = message.replaceAll("(pass=)[^,}]*", "$1****")
                .replaceAll("(user=)[^,}]*", "$1****");
        return sanitizedString;
    }
```
```suggestion
       logger.info(STR."Scheduling command \{sanitizedLogString(command.toString())} in \{delay.toMillis()} ms");
```
Which results in this being logged:
```
2024-07-21 10:39:39 INFO udpapi2.UdpApi scheduleCommand Scheduling command Command(action=AUTH, identifier=auth, tag=null, needsLogin=false, parameters={client=aniaddcli, clientver=4, comp=1, enc=UTF8, nat=1, pass=****, protover=3, user=****}, queryId=1) in 0 ms
```

File: src/main/java/udpapi2/UdpApi.java [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+                        return;
+                    }
+                }
+            }
+        }
+        if (commandCallbacks.containsKey(command.getClass())) {
+            //noinspection unchecked
+            commandCallbacks.get(command.getClass()).invoke(query);
+        } else {
+            logger.warning(STR."Unhandled query reply: \{query.toString()}");
```
Comment (runecalico) [inline src/main/java/udpapi2/UdpApi.java]: Not sure if you expect it to normally log this warning, even if it seems to do it only once (in a short test of 5 files)
```
2024-07-21 10:38:26 FINE processing.Mod_EpProcessing processEps Processing file Z:\xxx\anidbclitest1\A Silent Voice [anidb-10937]\A Silent Voice [BlurayDesuYo][BluRay][1920x1080][HEVC][7273A66D].mkv with Id 0
2024-07-21 10:39:39 FINE processing.Mod_EpProcessing onHashComputed File Z:\xxx\anidbclitest1\A Silent Voice [anidb-10937]\A Silent Voice [BlurayDesuYo][BluRay][1920x1080][HEVC][7273A66D].mkv with Id 0 has been hashed
2024-07-21 10:39:39 INFO udpapi2.UdpApi scheduleCommand Scheduling command Command(action=AUTH, identifier=auth, tag=null, needsLogin=false, parameters={client=aniaddcli, clientver=4, comp=1, enc=UTF8, nat=1, pass=****, protover=3, user=****}, queryId=1) in 0 ms
2024-07-21 10:39:39 FINE udpapi2.Send run Send command auth
2024-07-21 10:39:39 FINE processing.Mod_EpProcessing onHashComputed Requested Data for file with Id 0: SendFile: true, SendML: false
2024-07-21 10:39:39 FINE processing.Mod_EpProcessing processEps Processing file Z:\xxx\anidbclitest1\A Silent Voice [anidb-10937]\A Silent Voice [Okay-Subs][BluRay][1920x1036][HEVC][6457E1C3].mkv with Id 1
2024-07-21 10:39:39 FINE udpapi2.ParseReply parseReply Reply: auth-1 200 RKbYk xxxx.xxx.xxx.xxx:3333 LOGIN ACCEPTED \n
2024-07-21 10:39:39 INFO udpapi2.UdpApi handleQueryReply Logged in with session RKbYk
2024-07-21 10:39:39 WARNING udpapi2.UdpApi handleQueryReply Unhandled query reply: Query(command=Command(action=AUTH, identifier=auth, tag=null, needsLogin=false, parameters={client=aniaddcli, clientver=4, comp=1, enc=UTF8, nat=1, pass=****, protover=3, user=****}, queryId=1), reply=Reply(fullMessage=auth-1 200 RKbYk xxx.xxx.xxx.xxx:3333 LOGIN ACCEPTED
, fullTag=auth-1, replyStatus=LOGIN_ACCEPTED, queryId=null, responseData=[RKbYk, xxx.xxx.xxx.xxx:3333, LOGIN, ACCEPTED
]), sentAt=Sun Jul 21 10:39:39 CDT 2024, receivedAt=null, retries=0)
```

File: src/main/java/udpapi2/UdpApi.java (hunk 3 above)
Comment (Tyderion) [reply src/main/java/udpapi2/UdpApi.java]: oh woops.... that's a very good point 🙈 i'll implement that before merging

File: src/main/java/udpapi2/UdpApi.java (hunk 4 above)
Comment (Tyderion) [reply src/main/java/udpapi2/UdpApi.java]: yeah currently that's expected, i'll have to change that to make sure only actually undhandled replies trigger this. (login is handled by the API itself, all other commands must (should) be handled by the consumer of the api)

File: src/main/java/udpapi2/UdpApi.java (hunk 3 above)
Comment (Tyderion) [reply src/main/java/udpapi2/UdpApi.java]: ~~i've added a todo for now, will fix this week~~ 
fixed everywhere, log is now:
`2024-07-21 23:10:39 INFO udpapi.UdpApi scheduleCommand Scheduling command Command(action=AUTH, identifier=auth, tag=null, needsLogin=false, parameters={nat=1, comp=1, protover=3, pass=***, client=aniaddcli, clientver=4, enc=UTF8, user=***}, queryId=1) in 0 ms`

File: src/main/java/udpapi2/UdpApi.java (hunk 4 above)
Comment (Tyderion) [reply src/main/java/udpapi2/UdpApi.java]: ~~i've added a todo for now, will fix this week~~ 
Also fixed. And multiple callbacks for the same query/replystatus are supported as well

File: src/main/java/fileprocessor/FindFiles.java (hunk 1 above)
Comment (Tyderion) [reply src/main/java/fileprocessor/FindFiles.java:33]: i think you had an older version, but tell me if that's wrong.
I forgot to check for directories in `shouldScrapeFile` (line 43). 
It currently should not return any directories, but if that doesn't work with SMB shares, let me know and i'll fix it

File: src/main/java/aniAdd/startup/validation/validators/ValidationAnnotation.java [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package aniAdd.startup.validation.validators;
+
+public interface ValidationAnnotation {
```
Comment (Tyderion) [inline src/main/java/aniAdd/startup/validation/validators/ValidationAnnotation.java]: unused, delete!

**Comment by runecalico** [comment]
it looks like a change to the renaming methods no longer properly handles situations where the filename (including path) is "too long" and needs to be truncated (which the old code looks to have automatically truncated)
```
2024-07-25 10:07:22 FINE udpapi.ParseReply parseReply Reply: file:0-0 220 FILE \n 3404809|18164|274088|15782|0||0|0|7ae1ffc2|10|very
high|www|(HE-)AAC|HEVC|1920x1080|japanese|english'french'german'italian'spanish'spanish'portuguese'russian'arabic|1430|1704585600|Loop 7-kaime no Akuyaku Reijou wa, Moto Tekikoku de Jiyuukimama na Hanayome Seikatsu o
Mankitsu Suru - 01 - The Fiance Who Killed Me - [Judas](7ae1ffc2).mkv||12|12|2024-2024|TV Series||Loop 7-kaime no Akuyaku Reijou wa, Moto Tekikoku de Jiyuukimama na Hanayome Seikatsu o Mankitsu Suru|???7????????,
???????????????????|7th Time Loop: The Villainess Enjoys a Carefree Life Married to Her Worst Enemy!|???7????????, ???????????????????'7th Time Loop: The Villainess Enjoys a Carefree Life Married to Her Worst
Enemy!'?????????????????????????'??7??????????????????????|01|The Fiance Who Killed Me|Watashi o Koroshita Konyakusha|????????|Judas|Judas \n
2024-07-25 10:07:22 FINE processing.EpisodeProcessing aniDBInfoReply Got DB Info for file Z:\1-InboundRaw.PreSort\anidbclitest1\7th Time Loop The Villainess Enjoys a Carefree Life Married to Her Worst Enemy!
[anidb-18164]\7th Time Loop The Villainess Enjoys a Carefree Life Married to Her Worst Enemy! - 01 - The Fiance Who Killed.mkv with Id 0
java.lang.NullPointerException
        at java.base/java.util.Objects.requireNonNull(Objects.java:233)
        at java.base/java.util.stream.Collectors.lambda$uniqKeysMapAccumulator$1(Collectors.java:180)
        at java.base/java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
        at java.base/java.util.HashMap$EntrySpliterator.forEachRemaining(HashMap.java:1858)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
        at processing.tagsystem.TagSystem.Evaluate(TagSystem.java:18)
        at processing.FileRenamer.getPathFromTagSystem(FileRenamer.java:172)
        at processing.FileRenamer.getTargetFolder(FileRenamer.java:136)
        at processing.FileRenamer.renameFile(FileRenamer.java:29)
        at processing.EpisodeProcessing.finalProcessing(EpisodeProcessing.java:227)
        at processing.EpisodeProcessing.aniDBInfoReply(EpisodeProcessing.java:163)
        at udpapi.UdpApi.lambda$handleQueryReply$3(UdpApi.java:237)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
        at udpapi.UdpApi.handleQueryReply(UdpApi.java:237)
        at udpapi.UdpApi.addReply(UdpApi.java:223)
        at udpapi.ParseReply.parseReply(ParseReply.java:67)
        at udpapi.ParseReply.run(ParseReply.java:23)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
2024-07-25 10:07:22 SEVERE processing.FileRenamer renameFile Renaming failed for File Z:\1-InboundRaw.PreSort\anidbclitest1\7th Time Loop The Villainess Enjoys a Carefree Life Married to Her Worst Enemy!
```

**Comment by Tyderion** [comment]
oh shit. thanks for the headsup <3. 
it should still truncate (I still have the 240 char limit in there), but it looks like some value of the repsonse gets converted to null (probably some of those questionmarks in the log) which the toMap collector doesn't like (looks like there even is an java bug for that https://bugs.openjdk.org/browse/JDK-8148463).  i will try to fix it (and also find a similar file somewhere so i can verify myself).
I really should've written some tests, but it's really hard to find all the edge cases.

**Comment by runecalico** [comment]
> oh shit. thanks for the headsup <3. it should still truncate (I still have the 240 char limit in there), but it looks like some value of the repsonse gets converted to null (probably some of those questionmarks in the log) which the toMap collector doesn't like (looks like there even is an java bug for that https://bugs.openjdk.org/browse/JDK-8148463). i will try to fix it (and also find a similar file somewhere so i can verify myself). I really should've written some tests, but it's really hard to find all the edge cases.

I don't know if it will be helpful, but the previous rename code didn't have a problem with it, I originally renamed that file using one of the commits from this PR (which is when I requested the exit on ban) ..

Well, I guess that's what I'll end up doing for you. Finding a bunch of edge cases.
I've got easily 50000-100000 files I was running this against .. 

In case it's related to the tagging here is my tagging config ..
```
tagSystemCode: |-
  BaseTVShowPath:="Z:\2-WorkingImport-AniAdd\releases\"
  BaseEnglishTVShowPath:="Z:\2-WorkingImport-English-AniAdd\releases\"
  BaseMoviePath:="Z:\2-WorkingImport-AniAdd\movies\"
  BaseEnglishMoviePath:="Z:\2-WorkingImport-English-AniAdd\movies\"
  ShowTitle:=[%ATe%, %ATr%, %ATk%]
  EpisodeTitle:=[%ETe%, %ETr%, %ETk%]
  ShowTitle:=$repl(%ShowTitle%, "\?|\s*\.+$", "_")
  ShowTitle:=$repl(%ShowTitle%, '[\\":/*|<>?]', " ")
  EpisodeTitle:=$repl(%EpisodeTitle%, '[\\":/*|<>?]', " ")
  ShowTitle:=$repl(%ShowTitle%, "\s+", " ")
  EpisodeTitle:=$repl(%EpisodeTitle%, "\s+", " ")
  ShowTitle:=$repl(%ShowTitle%, "^\s|\s$", "")
  EpisodeTitle:=$repl(%EpisodeTitle%, "^\s|\s$", "")
  ShowTitle:=$repl(%ShowTitle%, "`", "'")
  EpisodeTitle:=$repl(%EpisodeTitle%, "`", "'")
  Trunc(str, len):=$repl(%str%, ".{" $len($repl(%str%, "(.?){" %len% "}$", "")) "}$", "")
  TruncEllipse(str, len):={$len(%str%) = $len($Trunc(%str%, %len%)) ? %str% : $Trunc(%str%, %len%) "…" }
  EpisodeTitle:=$TruncEllipse(%EpisodeTitle%, "32")
  Regular:=""
  Special:="S"
  Separator:=" - "
  SpecialEp:=$repl(%EpNo%, "[1234567890]", "")
  EpNo:=$repl(%EpNo%, "[SCTPO]", "")
  Ver:={%Ver% = "1" ? "" : "v" %Ver% }
  Pad:={%SpecialEp% ? "2" : $max($len(%EpHiNo%), $len(%EpCount%)) }
  Pad:={$match(%EpCount%, "0") ? $max("2", %Pad%) : %Pad% }
  EpNoPad:=$pad(%EpNo%, %Pad%, "0")
  EpNoPad:={%SpecialEp%       ? %EpNoPad%  :  %Regular%   %EpNoPad% }
  EpNoPad:={%SpecialEp% = "S" ? %Special%     %EpNoPad% : %EpNoPad% }
  EpNoPad:={%SpecialEp% = "C" ? %Special% "1" %EpNoPad% : %EpNoPad% }
  EpNoPad:={%SpecialEp% = "T" ? %Special% "2" %EpNoPad% : %EpNoPad% }
  EpNoPad:={%SpecialEp% = "P" ? %Special% "3" %EpNoPad% : %EpNoPad% }
  EpNoPad:={%SpecialEp% = "O" ? %Special% "4" %EpNoPad% : %EpNoPad% }
  EpNoFull:=%Separator% %EpNoPad% %Ver% %Separator%
  GT:="[" [%GTs%, %GTl%] "]"
  Src:="[" $repl(%Source%, "Blu-ray", "BluRay") "]"
  Cen:={%Cen% ? "[Cen]" : "" }
  Res:="[" %FVideoRes% "]"
  VCodec:="[" $repl(%FVCodec%,"H264/AVC","h264") "]"
  ACodec:="[" %FACodec% "]"
  CRC:="[" $uc(%FCrc%) "]"
  FileInfo:=" " %GT% %Src% %Cen% %Res% %VCodec% %CRC%
  MovieTypes:="Movie|OVA"
  IsMovie:={$match(%Type%, %MovieTypes%) ? {%EpCount% = "1" ? {%SpecialEp% ? "" : "1" } : "" } : "" }
  LangTypes:=".*([eE]nglish).*"
  IsEnglish:={$match(%FALng%, %LangTypes%) ? "1" : "" }
  MovieFileName:=%ShowTitle% %FileInfo%
  TVShowFileName:=%ShowTitle% %EpNoFull% %EpisodeTitle% %FileInfo%
  BasePath:={%IsMovie% ? { %IsEnglish% ? %BaseEnglishMoviePath% : %BaseMoviePath% } : { %IsEnglish% ? %BaseEnglishTVShowPath% : %BaseTVShowPath% } }
  FileName:={%IsMovie% ? %MovieFileName% : %TVShowFileName% }
  PathName:=%BasePath%  %ShowTitle% " [anidb-"%AId%"]"
```

**Comment by Tyderion** [comment]
oh wow, that's a sizeable collection. i only have around 8k files to test with.
I'm really glad that you are supporting this and helping me iron out the bugs that i introduced, thank you!


I did download some version of the Loop 7 anime just now and it seemed worked with the code that failed for you, but in my log the japanese letters did not get converted to questionmarks, i.e.
`2024-07-25 21:35:24 FINE udpapi.ParseReply parseReply Reply: file:0-0 220 FILE \n 3405188|18164|274088|11111|0||1|4096|60587c46|10|very high|www|OPUS|HEVC|1920x1080|japanese|english'portuguese'spanish'spanish'arabic'french'german'italian'russian|1430|1704585600|Loop 7-kaime no Akuyaku Reijou wa, Moto Tekikoku de Jiyuukimama na Hanayome Seikatsu o Mankitsu Suru - 01 - The Fiance Who Killed Me - [Hi10](60587c46).mkv||12|12|2024-2024|TV Series||Loop 7-kaime no Akuyaku Reijou wa, Moto Tekikoku de Jiyuukimama na Hanayome Seikatsu o Mankitsu Suru|ループ7回目の悪役令嬢は, 元敵国で自由気ままな花嫁生活を満喫する|7th Time Loop: The Villainess Enjoys a Carefree Life Married to Her Worst Enemy!|ループ7回目の悪役令嬢は, 元敵国で自由気ままな花嫁生活を満喫する'7th Time Loop: The Villainess Enjoys a Carefree Life Married to Her Worst Enemy!'輪迴七次的惡役千金，在前敵國享受隨心所欲的新婚生活'轮回7次的恶役千金，在前敌国享受随心所欲的新婚生活|01|The Fiance Who Killed Me|Watashi o Koroshita Konyakusha|私を殺した婚約者|Hi10 Anime|Hi10 \n 
`, so my next guess would be an encoding problem.
Could you try that single file with the newest commit? At least the exception should (hopefully lol) be different.


And it did shorten the name for me:
`Loop 7-kaime no Akuyaku Reijou wa, Moto Tekikoku de Jiyuukimama na Hanayome Seikatsu o Mankitsu Suru S01E01 T.mkv` was the final name, barely even containing the episode number. 
I think the truncating needs to be smarter and make sure to always include the episode numbers.
I'll have to think about how to do that for all cases though.

I will check with your tagsystem, maybe that's the difference. (sadly no). I think this still points to an encoding problem (you should not have so many `?` in your log)
With your tagsystem it renamed the file i got to
`7th Time Loop The Villainess Enjoys a Carefree Life Married to Her Worst Enemy! - 01 - The Fiance Who Killed Me [Hi1.mkv`

Obviously, the file i have is not the same, so it could be something even more difficult to test lol

**Comment by Tyderion** [comment]
never mind, i saw the file id in the log so i can test :)
Followup: well that didn't help, it still worked the same as with my file. I even tried on a seperate computer (to make sure it's not some config i changed on mine), but it still worked

Do you use win10 or win11?

**Comment by runecalico** [comment]
I just compiled commit 313fadf, and can confirm at least for that anime it renames it now and doesn't throw an error :)

and I'm currently running the jar on Windows Server 2019 (I think), at some point I'll switch to Server 2022 .. I'm working on some automation that will run it ..

**Comment by Tyderion** [comment]
oh that's good.

I plan to implement one more fix:
At the moment, if `--exit-on-ban` is false, it loses the command that just ended in an error if you get banned (or any other fatal error like anidb being down), so i want to make sure to requeue that missed command.

After that I plan on merging this and releasing a release candidate `4.0.0-rc1` (or something similar) so that future fixes/features can be done in smaller PRs (this one got out of hand too quickly 😂 )

I will also be adding a few more issues for possible future development, you're welcome to add or comment on any of them of course 😄

**Comment by runecalico** [comment]
Seems `--exit-on-ban` still needs some work, but it's _close_. This is from commit 313fadf7747273dd3010f413293822dea5a7c6a5
```
2024-07-26 22:19:23 FINE processing.EpisodeProcessing finalProcessing File Z:\1-InitialSort\releases\Shen Wu Tianzun [anidb-17085]\The Legend of Sky Lord Episode 088.mp4 with Id 12479 done
2024-07-26 22:19:23 INFO udpapi.UdpApi scheduleCommand Scheduling command Command(action=FILE, identifier=file, tag=12480, needsLogin=true, parameters={fmask=7F0AEAE920, amask=F2F0F0C0, size=226731674,
ed2k=933f69641bbcbda111eae609a2adfa8b}, queryId=12481) in 3797 ms
2024-07-26 22:19:27 FINE udpapi.Send run Send command file
2024-07-26 22:19:27 FINE udpapi.ParseReply parseReply Reply: 555 BANNED \n Leech \n
2024-07-26 22:19:27 WARNING udpapi.ParseReply run Parse Error: java.lang.NullPointerException: Cannot invoke "Object.hashCode()" because "key" is null
java.lang.NullPointerException: Cannot invoke "Object.hashCode()" because "key" is null
        at java.base/java.util.concurrent.ConcurrentHashMap.get(ConcurrentHashMap.java:936)
        at java.base/java.util.concurrent.ConcurrentHashMap.containsKey(ConcurrentHashMap.java:964)
        at udpapi.UdpApi.addReply(UdpApi.java:208)
        at udpapi.ParseReply.parseReply(ParseReply.java:67)
        at udpapi.ParseReply.run(ParseReply.java:23)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
```
and then it hangs (I think). I'm running this in a pwsh script and it should exit the pwsh script if the process exits with a non-zero rc .. but the script was still running, so maybe the command didn't exit after the exception?

**Comment by Tyderion** [comment]
gah, thanks for that. it shouldn't even try to access the hashmap at all if it is a ban/fatal reply. i'll fix that today. thanks for the stacktrace, that should make it easy to fix :)
or not, i'll need to look into that a little bit more than expected
Okay, i see why it fails. should have a fix out today. stupid mistake 🙈 
When a command is successfuly, it's format is `some-tag STATUS name data` e.g. `file:7-8 220 FILE [...]`, but if it's banned, it's just `STATUS name data`, e.g. `555 BANNED [...]`
